fletcher/base.py

```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
             When `indices` contains negative values other than ``-1``
             and `allow_fill` is True.
 
-        Notes
-        -----
-        ExtensionArray.take is called by ``Series.__getitem__``, ``.loc``,
-        ``iloc``, when `indices` is a sequence of values. Additionally,
-        it's called by :meth:`Series.reindex`, or any other method
-        that causes realignemnt, with a `fill_value`.
-
         See Also
         --------
         numpy.take
         pandas.api.extensions.take
         """
-        from pandas.core.algorithms import take
+        if isinstance(indices, list):
+            indices = np.asarray(indices)
+            if not np.any(indices < 0):
+                allow_fill = False
+        if isinstance(indices,slice) and allow_fill:
+            raise ValueError('allow_fill cannot be used with slice')
+
+        if not allow_fill:
+            indices = self._indices_to_numpy_array(indices)
+            if self.data.num_chunks==1:
+                return(FletcherArray(self.data.chunk(0).take(pa.array(indices))))
+
+            lengths = np.fromiter(map(len, self.data.chunks), dtype=np.int64)
+            cum_lengths = lengths.cumsum()
+
+            bins = np.searchsorted(cum_lengths, indices, side='right')  #one might want to parallel this operation
```

Comment:
given `bins` and `lengths` we could better chose an optimal path between `self.combine_array().take(...)` and the scenario below !